### PR TITLE
[pigeon] Implement Screaming Snake Case Conversion for Kotlin Enum Cases

### PR DESCRIPTION
**Pull request by erdzan12**
This pull request addresses issue https://github.com/flutter/flutter/issues/140938 in the Pigeon package, related to the naming convention of Kotlin enum cases generated from lower camel case Dart enums. The current implementation concatenates the enum cases in uppercase, deviating from the Kotlin naming convention, specifically when dealing with multi-word names.

Changes

- Kotlin Enum Generation: Modified the writeEnum function in the Pigeon package to ensure the Kotlin generator produces enum cases in SCREAMING_SNAKE_CASE. This adheres to the Kotlin coding conventions and allows a consistent cross-platform enum naming convention across Dart, Kotlin, and Swift.
- Regex Handling: Enhanced the regex pattern to correctly transform lower camel case names to screaming snake case, considering edge cases involving numbers and special characters.
- Testing: Updated the Dart unit tests to include cases for validating the correct transformation of multi-word and complex enum names from lower camel case to screaming snake case.

## Pre-launch Checklist

- [x] I read the [Contributor Guide] and followed the process outlined there for submitting PRs.
- [x] I read the [Tree Hygiene] wiki page, which explains my responsibilities.
- [x] I read and followed the [relevant style guides] and ran the auto-formatter. (Unlike the flutter/flutter repo, the flutter/packages repo does use `dart format`.)
- [x] I signed the [CLA].
- [x] The title of the PR starts with the name of the package surrounded by square brackets, e.g. `[shared_preferences]`
- [x] I listed at least one issue that this PR fixes in the description above.
- [x] I updated `pubspec.yaml` with an appropriate new version according to the [pub versioning philosophy], or this PR is [exempt from version changes].
- [x] I updated `CHANGELOG.md` to add a description of the change, [following repository CHANGELOG style].
- [x] I updated/added relevant documentation (doc comments with `///`).
- [x] I added new tests to check the change I am making, or this PR is [test-exempt].
- [x] All existing and new tests are passing.

If you need help, consider asking for advice on the #hackers-new channel on [Discord].

[Contributor Guide]: https://github.com/flutter/packages/blob/main/CONTRIBUTING.md
[Tree Hygiene]: https://github.com/flutter/flutter/wiki/Tree-hygiene
[relevant style guides]: https://github.com/flutter/packages/blob/main/CONTRIBUTING.md#style
[CLA]: https://cla.developers.google.com/
[flutter/tests]: https://github.com/flutter/tests
[breaking change policy]: https://github.com/flutter/flutter/wiki/Tree-hygiene#handling-breaking-changes
[Discord]: https://github.com/flutter/flutter/wiki/Chat
[pub versioning philosophy]: https://dart.dev/tools/pub/versioning
[exempt from version changes]: https://github.com/flutter/flutter/wiki/Contributing-to-Plugins-and-Packages#version-and-changelog-updates
[following repository CHANGELOG style]: https://github.com/flutter/flutter/wiki/Contributing-to-Plugins-and-Packages#changelog-style
[test-exempt]: https://github.com/flutter/flutter/wiki/Tree-hygiene#tests

### CODE REVIEW

File: packages/pigeon/CHANGELOG.md
Comment (tarrinneal) [inline packages/pigeon/CHANGELOG.md:7]: Any breaking changes will need to have updated integration tests as well. 

If you run the generate tool, there should then be some errors in the kotlin code for the tests. 

Let me know if you need help with that.

**Review by tarrinneal — CHANGES_REQUESTED** [review]
I think this is a good change, it will need updated integration tests as well though.

Thanks for putting this together.

File: packages/pigeon/test/kotlin_generator_test.dart
Comment (erdzan12) [reply packages/pigeon/test/kotlin_generator_test.dart]: makes sense

File: packages/pigeon/CHANGELOG.md
Comment (erdzan12) [reply packages/pigeon/CHANGELOG.md:7]: Do you think it is ok like this, to have a new enum value, the `SnakeCase` to test the functionality for Kotlin, or should we update the existing enums, and thus the tests in the other places.

Thus, is using
```
  /// This comment is to test enum member (Pending) documentation comments.
  Pending,

  /// This comment is to test enum member (Success) documentation comments.
  Success,

  /// This comment is to test enum member (Error) documentation comments.
  Error,
  ```
  Some convention, or can it be changed?

Comment (tarrinneal) [reply packages/pigeon/CHANGELOG.md:7]: probably best to just change one of these, if you're willing to make those changes on all platforms.

Comment (tarrinneal) [reply packages/pigeon/CHANGELOG.md:7]: I think the change you made is a good one, but perhaps also add a multi-word enum to `AnEnum` in the core_tests pigeon as well?

**Comment by tarrinneal** [comment]
This shouldn't be passing tests as is, I'll go ahead and generate the new files, format them, and upload them to this pr (assuming the branch isn't preventing that).

@stuartmorgan for tests not failing here.

**Comment by erdzan12** [comment]
I was hesitant regarding the generated files, wanted to check if they would be generated when running the tests here on the CI, as I saw the generated files on my end had lots of changes but only regarding the formatting + a couple of new things regarding the additions

**Comment by tarrinneal** [comment]
> I was hesitant regarding the generated files, wanted to check if they would be generated when running the tests here on the CI, as I saw the generated files on my end had lots of changes but only regarding the formatting + a couple of new things regarding the additions

always good to run the formatting tool after generation to clean things up a bit: https://github.com/flutter/plugins/blob/master_archive/script/tool/README.md#format-code

**Comment by tarrinneal** [comment]
There should also me a few integration test in `packages/pigeon/platform_tests/shared_test_plugin_code/lib/integration_tests.dart` that echo's an enum, it would be good to add one of each that sends one of the new members across as well.

**Comment by tarrinneal** [comment]
@stuartmorgan for second review

**Comment by tarrinneal** [comment]
Sorry for the force push, I accidentally sent up a file I didn't mean to, had to revert that.

**Review by tarrinneal — APPROVED** [review]
As long as the tests pass, this lgtm

**Comment by stuartmorgan** [comment]
> @stuartmorgan for tests not failing here.

What tests weren't failing that should have been, specifically? There was a Flutter-wide issue where test failures didn't turn the step red, so we can inspect the output manually.

**Review by stuartmorgan — APPROVED** [review]
LGTM

**Comment by stuartmorgan** [comment]
(Holding off on autosubmit until the test question is resolved in case there's still something going on here that would result in us not seeing issues with this PR in CI.)

**Comment by tarrinneal** [comment]
> (Holding off on autosubmit until the test question is resolved in case there's still something going on here that would result in us not seeing issues with this PR in CI.)

wasn't related to the code itself. There was no updated generated files, but it didn't fail the formatting test.

**Pull request merged**
